argon2/tests/kat.rs

```diff
@@ -288,3 +290,572 @@ fn argon2id_v0x13() {
+    Version::V0x10,
+    2,
+    20,
+    1,
+    b"password",
+    b"somesalt",
+    hex!("9690ec55d28d3ed32562f2e73ea62b02b018757643a2ae6e79528459de8106e9"),
+    "$argon2i$m=1048576,t=2,p=1$c29tZXNhbHQ$lpDsVdKNPtMlYvLnPqYrArAYdXZDoq5ueVKEWd6BBuk",
+    Algorithm::Argon2i
+);
+
+testcase_good!(
+    reference_argon2i_v0x10_2_18_1,
+    Version::V0x10,
+    2,
+    18,
+    1,
+    b"password",
+    b"somesalt",
+    hex!("3e689aaa3d28a77cf2bc72a51ac53166761751182f1ee292e3f677a7da4c2467"),
+    "$argon2i$m=262144,t=2,p=1$c29tZXNhbHQ$Pmiaqj0op3zyvHKlGsUxZnYXURgvHuKS4/Z3p9pMJGc",
+    Algorithm::Argon2i
+);
+
+testcase_good!(
+    reference_argon2i_v0x10_2_8_1,
+    Version::V0x10,
+    2,
+    8,
+    1,
+    b"password",
+    b"somesalt",
+    hex!("fd4dd83d762c49bdeaf57c47bdcd0c2f1babf863fdeb490df63ede9975fccf06"),
+    "$argon2i$m=256,t=2,p=1$c29tZXNhbHQ$/U3YPXYsSb3q9XxHvc0MLxur+GP960kN9j7emXX8zwY",
+    Algorithm::Argon2i
+);
+
+testcase_good!(
+    reference_argon2i_v0x10_2_8_2,
+    Version::V0x10,
+    2,
+    8,
+    2,
+    b"password",
+    b"somesalt",
+    hex!("b6c11560a6a9d61eac706b79a2f97d68b4463aa3ad87e00c07e2b01e90c564fb"),
+    "$argon2i$m=256,t=2,p=2$c29tZXNhbHQ$tsEVYKap1h6scGt5ovl9aLRGOqOth+AMB+KwHpDFZPs",
+    Algorithm::Argon2i
+);
+
+testcase_good!(
+    reference_argon2i_v0x10_1_16_1,
+    Version::V0x10,
+    1,
+    16,
+    1,
+    b"password",
+    b"somesalt",
+    hex!("81630552b8f3b1f48cdb1992c4c678643d490b2b5eb4ff6c4b3438b5621724b2"),
+    "$argon2i$m=65536,t=1,p=1$c29tZXNhbHQ$gWMFUrjzsfSM2xmSxMZ4ZD1JCytetP9sSzQ4tWIXJLI",
+    Algorithm::Argon2i
+);
+
+testcase_good!(
+    reference_argon2i_v0x10_4_16_1,
+    Version::V0x10,
+    4,
+    16,
+    1,
+    b"password",
+    b"somesalt",
+    hex!("f212f01615e6eb5d74734dc3ef40ade2d51d052468d8c69440a3a1f2c1c2847b"),
+    "$argon2i$m=65536,t=4,p=1$c29tZXNhbHQ$8hLwFhXm6110c03D70Ct4tUdBSRo2MaUQKOh8sHChHs",
+    Algorithm::Argon2i
+);
+
+testcase_good!(
+    reference_argon2i_v0x10_2_16_1_differentpassword,
+    Version::V0x10,
+    2,
+    16,
+    1,
+    b"differentpassword",
+    b"somesalt",
+    hex!("e9c902074b6754531a3a0be519e5baf404b30ce69b3f01ac3bf21229960109a3"),
+    "$argon2i$m=65536,t=2,p=1$c29tZXNhbHQ$6ckCB0tnVFMaOgvlGeW69ASzDOabPwGsO/ISKZYBCaM",
+    Algorithm::Argon2i
+);
+
+testcase_good!(
+    reference_argon2i_v0x10_2_16_1_diffsalt,
+    Version::V0x10,
+    2,
+    16,
+    1,
+    b"password",
+    b"diffsalt",
+    hex!("79a103b90fe8aef8570cb31fc8b22259778916f8336b7bdac3892569d4f1c497"),
+    "$argon2i$m=65536,t=2,p=1$ZGlmZnNhbHQ$eaEDuQ/orvhXDLMfyLIiWXeJFvgza3vaw4kladTxxJc",
+    Algorithm::Argon2i
+);
+
+/* Argon2i V0x10: Error state tests */
+
+#[test]
+fn reference_argon2i_invalid_encoding_1() {
+    /* Handle an invalid encoding correctly (it is missing a $) */
+    let hash = PasswordHash::new(
+        "$argon2i$m=65536,t=2,p=1c29tZXNhbHQ$9sTbSlTio3Biev89thdrlKKiCaYsjjYVJxGAL3swxpQ",
+    )
+    .unwrap();
```

Comment:
A quick note that this isn't expected to immediately fail re: your comment here:

> But argon doesn't seem to complain (even from the p=1c29tZXNhbHQ !)

`PasswordHash` comes from the `password-hash` crate and doesn't eagerly validate the parameters for a particular algorithm:

https://docs.rs/password-hash/0.3.2/password_hash/struct.PasswordHash.html

The parameters can be validated by extracting them using the `TryFrom<&PasswordHash>` impl on `Params`: 

https://docs.rs/argon2/0.3.1/argon2/struct.Params.html#impl-TryFrom%3C%26%27a%20PasswordHash%3C%27a%3E%3E